streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -63,87 +78,122 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient,
+                } catch (final InterruptedException fatalException) {
+                    final String error = "Could not create internal topics.";
+                    log.error(error, fatalException);
+                    throw new StreamsException(error, fatalException);
+                }
+            }
+        }
+    }
+
+    private void verifyIfTopicExistsWithCorrectNumberOfPartitions(final String topic,
+                                                                  final InternalTopicConfig expectedNumberOfPartitions) {
+        final Map<String, Integer> existingTopicPartitions = getNumPartitions(Collections.singleton(topic));
+        if (validateTopicPartitions(Collections.singleton(expectedNumberOfPartitions), existingTopicPartitions).size() != 0) {
+            final String errorMsg = String.format("Could neither create topic %s nor receive any metadata about it.", topic);
+            log.error(errorMsg);
+            throw new StreamsException(errorMsg);
         }
-        throw new StreamsException("Could not create internal topics.");
     }
 
     /**
      * Get the number of partitions for the given topics
      */
     public Map<String, Integer> getNumPartitions(final Set<String> topics) {
-        for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
-            try {
-                final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
-                final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
-                existingTopicPartitions.keySet().retainAll(topics);
-
-                return existingTopicPartitions;
-            } catch (StreamsException ex) {
-                log.warn("Could not get number of partitions: {} Retry #{}", ex.getMessage(), i);
+        return getNumPartitions(topics, false);
+    }
+
+    private Map<String, Integer> getNumPartitions(final Set<String> topics,
+                                                  final boolean bestEffort) {
+        try {
+            final DescribeTopicsResult describeTopicsResult = adminClient.describeTopics(topics);
+            final Map<String, KafkaFuture<TopicDescription>> futures = describeTopicsResult.values();
+
+            final Map<String, Integer> existingTopicPartitions = new HashMap<>();
+            for (final Map.Entry<String, KafkaFuture<TopicDescription>> topicFuture : futures.entrySet()) {
+                try {
+                    final TopicDescription topicDescription = topicFuture.getValue().get();
+                    existingTopicPartitions.put(
+                        topicFuture.getKey(),
+                        topicDescription.partitions().size());
+                } catch (final ExecutionException topicDoesNotExist) {
+                    if (bestEffort) {
+                        log.warn("Could not get number of partitions for topic {}. Topic might not exists: {}",
```

Comment:
Isn't this more likely to happen in practice? Do we want to produce this as WARN? I felt making INFO or even DEBUG is better.